### DISCUSSION TITLE
Discrepancies in data for Generated Solar

### DISCUSSION THREAD

**Discussion by Captain0Kangaroo** [nathanmarlor/foxess_modbus#104] [General]
I've been using the Chesnut Modbus for the last 9 months or so  but wanted to take a look at yours @nathanmarlor but i've noticed some discrepancies in the solar generation between the two.

Yours is saying 6.4kw and the other is 6.2. Not massive i know (200w) but it was a poor day so on a good day this might be more? Need the sun to confirm that, but over time that 200w adds up. 

Is this because your system is reading the inverter more and collecting more data and is more accurate?

**Comment by canton7** [comment]
I think the StealthChestnut one uses an integration sensor for solar generation total / today? We found a register which we can read directly. So there's probably a small amount of error in the StealthChestnut one due to it only reading the current power once every however many seconds and having to interpolate between those points.

That said, there don't appear to be registers for total PV1/PV2, so we have to use integration sensors there too.

**Reply by Captain0Kangaroo** [reply]
Ah got it... 

One more thing which is not entirely the place for this question but not sure where else to ask...

I've been logging my generation by Day/ Week/ Month and Total with a template sensor and i'd really like to try and keep this data but i know when i change the sensor values this will reset to 0. Do you know if it's possible to manually set these numbers in HA so it would continue from where it left off? 

 
<img width="497" alt="Screen Shot 2023-03-31 at 19 48 05" src="https://user-images.githubusercontent.com/96144685/229204724-4d7fa38e-062c-4229-9be7-1c34682361bf.png">

**Reply by canton7** [reply]
Template sensors are easy. You just change the template and the next time that it creates a new value, it'll use the new template.

Presumably you're then feeding that into meters for the daily/etc? Those are fine too - they'll keep using the same source, and they'll never know that the template sensor was changed.

**Reply by canton7** [reply]
Although, if the source sensor decreases and the meters are set up to not expect a decrease, they'll reset true.

**Reply by Captain0Kangaroo** [reply]
It think most of the sensors had the same name so most things have remained (phew).
Just lost the main energy data but took screen shots so can look back if needed.

**Reply by Captain0Kangaroo** [reply]
> Template sensors are easy. You just change the template and the next time that it creates a new value, it'll use the new template.
> 
> Presumably you're then feeding that into meters for the daily/etc? Those are fine too - they'll keep using the same source, and they'll never know that the template sensor was changed.

@canton7 Sorry to bother you... I just can't seem to get my sensors working - i'm sure its easy but just cant figure it out.
I had a card which basically showed solar gen for Today/ Month/ Year / Total and each would reset accordinly. 
![Solar Gen Jan-March 23](https://user-images.githubusercontent.com/96144685/229273515-bc228679-a862-4ca4-be0c-2be2a9ad033c.jpg)

These were my sensors:

    pv_total_power:
          friendly_name: "PV Power now"
          unit_of_measurement: 'kW'
          device_class: power
          value_template: >
              {{ (states('sensor.pv1_power') | float(default=0) + states('sensor.pv2_power') | float(default=0) ) * 1 }}              
    pv_total_daily:
          friendly_name: "PV Energy daily"
          unit_of_measurement: 'kWh'
          device_class: energy
          value_template: >
              {{ ((states('sensor.pv1_daily_power') | float(default=0) + states('sensor.pv2_daily_power') | float(default=0) ) * 1 ) | round(2) }} 
    pv_total_monthly:
          friendly_name: "PV Energy Monthly"
          unit_of_measurement: 'kWh'
          device_class: energy
          value_template: >
              {{ ((states('sensor.pv1_monthly_power') | float(default=0) + states('sensor.pv2_monthly_power') | float(default=0) ) * 1 ) | round(2) }} 
    pv_total_yearly:
          friendly_name: "PV Energy Yearly"
          unit_of_measurement: 'kWh'
          device_class: energy
          value_template: >
              {{ ((states('sensor.pv1_yearly_power') | float(default=0) + states('sensor.pv2_yearly_power') | float(default=0) ) * 1 ) | round(2) }}
    pv_total_total:
          friendly_name: "PV Energy Total"
          unit_of_measurement: 'kWh'
          device_class: energy
          value_template: >
              {{ ((states('sensor.pv1_total_power') | float(default=0) + states('sensor.pv2_total_power') | float(default=0) ) * 1 ) | round(2) }}

But since the change over they have stopped working and like i say for the life of me cant get them working again. 
Any ideas what i'm missing?
Thanks

**Reply by canton7** [reply]
The sensors which those templates rely on - pv1_monthly_power etc - aren't provided by either integration. You must have created them yourself, probably as energy meters?

**Reply by Captain0Kangaroo** [reply]
Yes it was something I added in with the help of someone about a year ago. All this yaml stuff is a bit beyond me to be honest.  
Its just a running total which i assume relates to this:
  pv1_monthly_power:
    source: sensor.pv1_sum
    cycle: monthly
  pv2_monthly_power:
    source: sensor.pv2_sum
    cycle: monthly
  pv1_yearly_power:
    source: sensor.pv1_sum
    cycle: yearly
  pv2_yearly_power:
    source: sensor.pv2_sum
    cycle: yearly
  pv1_total_power:
    source: sensor.pv1_sum
  pv2_total_power:
    source: sensor.pv2_sum

I though by changing the sensor names to whats used for this integration it would start working but its not.

**Reply by canton7** [reply]
Those source sensors pv1_sum and pv2_sum are still the StealthChestnut sensors. The ones from this integration are called pv1_energy_total and pv2_energy_total.

**Reply by Captain0Kangaroo** [reply]
I've changed the sensors to this for example:

    pv_total_monthly:
          friendly_name: "PV Energy Monthly"
          unit_of_measurement: 'kWh'
          device_class: energy
          value_template: >
              {{ ((states('sensor.pv1_power_monthly') | float(default=0) + states('sensor.pv2_power_monthly') | float(default=0) ) * 1 ) | round(2) }} 

and changed the utility meter to:

  pv1_power_monthly:
    source: sensor.pv1_sum
    cycle: monthly
  pv2_power_monthly:
    source: sensor.pv2_sum
    cycle: monthly

Something is now coming through but not sure its right with the lack of sun right now :-)

**Reply by Captain0Kangaroo** [reply]
> Those source sensors pv1_sum and pv2_sum are still the StealthChestnut sensors. The ones from this integration are called pv1_energy_total and pv2_energy_total.

ah so i'd need to change them to pv1_energy_total and pv2_energy_total?

**Reply by canton7** [reply]
Yes. At that point they might reset, as the totals recorded by this integration are less than the totals recorded by the StealthChestnut intigration, as that one's been running for longer and has seen more generation.

**Reply by Captain0Kangaroo** [reply]
Thanks @canton7  I'll give that a try... Just need the sun now!

**Reply by nathanmarlor** [reply]
Do you have an update on this please @Captain0Kangaroo ?

**Reply by Captain0Kangaroo** [reply]
> Do you have an update on this please @Captain0Kangaroo ?

No not really.. I'm just using it 🤣
All good mate 👍
